### PR TITLE
Fixed image uploading. Both URL and a file

### PR DESCRIPTION
**Pull request by korzol**
Hello
Thank you! You did a great job!
iZettle updated their API documentation. At least new there is correct URL to send images to. 
Also some small changes like uuid in ImageClient was null. 
Also now they require image type even for url, so I had to implement it somehow

Have a good day

### CODE REVIEW

**Comment by LauLaman** [comment]
Great job! Can you also add in the changes in the tests?

**Comment by korzol** [comment]
Will be glad but first I need to learn how it works and where to make changes :)

**Comment by korzol** [comment]
Added image upload to unit tests. But testing fails if I run it as is. Complain is:

```
Fatal error: Trait 'LauLamanApps\IzettleApi\Tests\Unit\MockeryAssertionTrait' not found in ..izettle/vendor/laulamanapps/izettle-api/tests/Unit/GuzzleIzettleClientTest.php on line 24
```
Checked path and namespaces, even added 
```php
require_once 'MockeryAssertionTrait.php'; 
```
to GuzzleIzettleClientTest.php but seems nothing works.  

PHP 7.2.15.
iZettleApi has been installed with composer

Any thoughts ?

File: tests/Unit/Client/Image/ImageUrlUploadTest.php [hunk 1]
```diff
@@ -17,12 +17,25 @@ final class ImageUrlUploadTest extends TestCase
...
+            strtolower(array_values(array_slice(explode('.', $imageUrl), -1))[0])
+        ];
 
         $productImageUpload = new ImageUrlUpload($imageUrl);
 
         self::assertInstanceOf(ImageUrlUpload::class, $productImageUpload);
+        self::assertAttributeEquals($imageFormat, 'imageFormat', $productImageUpload);
         self::assertAttributeEquals($imageUrl, 'imageUrl', $productImageUpload);
-        self::assertSame(['imageUrl' => $imageUrl], json_decode($productImageUpload->getPostBodyData(), true));
+        // self::assertSame(['imageUrl' => $imageUrl], json_decode($productImageUpload->getPostBodyData(), true));
```
Comment (LauLaman) [inline tests/Unit/Client/Image/ImageUrlUploadTest.php]: Please remove dead code
```suggestion
```

File: tests/Unit/Client/Image/ImageUrlUploadTest.php [hunk 2]
```diff
@@ -17,12 +17,25 @@ final class ImageUrlUploadTest extends TestCase
...
+        self::assertSame(['imageFormat' => $imageFormat, 'imageData' => $imageData], json_decode($productImageUpload->getPostBodyData(), true));
```
Comment (LauLaman) [inline tests/Unit/Client/Image/ImageUrlUploadTest.php]: This should fix this test

```suggestion
        self::assertSame(['imageFormat' => $imageFormat, 'imageUrl' => $imageUrl], json_decode($productImageUpload->getPostBodyData(), true));
```

File: tests/Unit/Client/Image/ImageUrlUploadTest.php [hunk 3]
```diff
@@ -17,12 +17,25 @@ final class ImageUrlUploadTest extends TestCase
...
+        $allowedImageType = [
+            'gif'  => 'GIF',
+            'jpeg' => 'JPEG',
+            'jpg'  => 'JPEG',
+            'png'  => 'PNG',
+            'bmp'  => 'BMP',
+        ];
+
+        $imageUrl       = 'https://example.com/image.png';
+        $imageFormat    = $allowedImageType[
```
Comment (LauLaman) [inline tests/Unit/Client/Image/ImageUrlUploadTest.php]: Please make sure the code style is inline with the other files:
(alignment equal signs)

run `make cs-fix` to fix automatically

File: src/Client/Image/ImageFileUpload.php [hunk 4]
```diff
@@ -31,7 +31,22 @@ public function __construct(string $filename)
...
+        $this->imageData = $this->byteArray($filename);
+    }
+
+    /**
+     *  Generate byte array similar to the one in C# and Java
+     */
+    public function byteArray(string $filename): array
+    {
+        $data   = file_get_contents($filename);
+        $array  = array();
```
Comment (LauLaman) [inline src/Client/Image/ImageFileUpload.php]: Please make sure the code style is inline with the other files:
(Short array syntax)

run `make cs-fix` to fix automatically

File: src/Client/Image/ImageFileUpload.php [hunk 5]
```diff
@@ -31,7 +31,22 @@ public function __construct(string $filename)
     {
         $this->validateFile($filename);
         $this->imageFormat = self::ALLOWED_FILE_TYPES[exif_imagetype($filename)];
-        $this->imageData = iconv('UTF-8', 'UTF-8//IGNORE', utf8_encode(file_get_contents($filename)));
+        $this->imageData = $this->byteArray($filename);
+    }
+
+    /**
+     *  Generate byte array similar to the one in C# and Java
```
Comment (LauLaman) [inline src/Client/Image/ImageFileUpload.php]: Please remove unnecessary comment. It does not add any value.

File: src/Client/Image/ImageFileUpload.php [hunk 6]
```diff
@@ -31,7 +31,22 @@ public function __construct(string $filename)
     {
         $this->validateFile($filename);
         $this->imageFormat = self::ALLOWED_FILE_TYPES[exif_imagetype($filename)];
```
Comment (LauLaman) [inline src/Client/Image/ImageFileUpload.php]: Can you add `ext-exif` to composer.json as a required extension? i've missed this.

**Comment by LauLaman** [comment]
Hi @korzol 
Great job!

Can you tell me how you are running the tests? are you using the Make commands?

I've checkout your pr locally and have come to the following conclusion:
- `LauLamanApps\IzettleApi\Tests\Integration\Client\ImageClientTest:26` We need to enter a real url now. this will fix this tests. i've tried with `http://example.com/image.jpg`
- Same goes for `LauLamanApps\IzettleApi\Tests\Unit\Client\ImageClientTest:27`

If i run with the make command it should work. You can also check if you have mockery installed in your vendor directory.

i've also left you some more feedback.

**Comment by LauLaman** [comment]
Hi @korzol

Are you able to finish this PR?

**Comment by chatlumo** [comment]
Hello @korzol,

Is there any news about this PR ?

**Comment by LauLaman** [comment]
Closed due to no response. i might fix this miself later on

**Pull request closed without merging**